src/playbin.js

```diff
@@ -102,63 +132,81 @@ var Playbin = GObject.registerClass({
         if (width_ok && height_ok) {
             this._video_width = width;
             this._video_height = height;
-            this._video_widget.show_now();
+            this._updateCropArea(this.get_allocation());
         }
     }
 
-    _playbackDone () {
+    _onEndOfStream () {
         this.emit('done');
         this._playbin.set_state(Gst.State.NULL);
         this.remove(this._video_widget);
-        this.uri = null;
+        this._uri = null;
         this._playbin = null;
         this._videocrop = null;
         this._video_widget = null;
+        this._video_width = 0;
+        this._video_height = 0;
+    }
+
+    _onStateChanged (msg) {
+        if (this._started)
+            return;
+
+        const [, newstate] = msg.parse_state_changed();
+
+        if (newstate === Gst.State.PLAYING) {
+            this.set_visible_child(this._video_widget);
+            this._started = true;
+        }
     }
 
     _bus_watch (bus, msg) {
         if (msg.type === Gst.MessageType.EOS) {
-            this._playbackDone();
+            this._onEndOfStream(msg);
             return GLib.SOURCE_REMOVE;
         } else if (msg.type === Gst.MessageType.STREAMS_SELECTED) {
-            this._updateSizeFromStreamCollection(msg.parse_streams_selected());
+            this._onStreamsSelected(msg);
         } else if (msg.type === Gst.MessageType.STATE_CHANGED) {
-            const [, newstate] = msg.parse_state_changed();
-            this._state = newstate;
+            this._onStateChanged(msg);
         }
 
         return GLib.SOURCE_CONTINUE;
     }
 
-    setBackground (url) {
-        try {
-            this._css_provider.load_from_data(`
-playbin.locked {
-    background: url('${url}') no-repeat center;
-}`
-            );
-        } catch (e) {
-            logError(e, 'Failed to load css');
-        }
+    get uri () {
+        return this._uri;
     }
 
-    play () {
-        if (!this.uri)
+    set uri (value) {
+        if (!value)
             return;
 
         this._ensurePlaybin();
+        this._uri = value;
+        this._playbin.uri = this._uri;
+        this._playbin.set_state(Gst.State.PAUSED);
+    }
+
+    play () {
+        if (!this._uri)
+            return;
 
         this._started = false;
-        this._playbin.uri = this.uri;
+        this._playbin.uri = this._uri;
         this._playbin.set_state(Gst.State.PLAYING);
     }
 
+    set background (url) {
+        this._button.background = url;
+    }
+
     set locked(value) {
-        const style = this.get_style_context();
-        if (value)
-            style.add_class('locked');
-        else
-            style.remove_class('locked');
+        if (value) {
+            this._button.show();
+            this.set_visible_child(this._button);
+        } else if (!this._uri) {
```

Comment:
I am not sure why this check needs to be here... Feels to me like it'd belong in whoever calls `locked = false`. I would like to keep this function "dumb" and just set the locked value it's passed, without state checks.